[inductor] coordinate descent tuning upon max-autotune

shunting314 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99403
* __->__ #97203
* #99216

Command to run max autotune baseline:
```
TORCHINDUCTOR_MAX_AUTOTUNE=1 time python benchmarks/dynamo/torchbench.py --backend inductor --amp --performance --only ${MODEL_NAME} --training --batch-size-file $(realpath benchmarks/dynamo/torchbench_models_list.txt)
```

Command to do coordinate descent autotuning: 
```
TORCHINDUCTOR_COORDINATE_DESCENT_TUNING=1 TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_shunting_coordesc TORCHINDUCTOR_PERSISTENT_REDUCTIONS=0 TORCHINDUCTOR_MAX_AUTOTUNE=1 time python benchmarks/dynamo/torchbench.py --backend inductor --amp --performance --only ${MODEL_NAME} --training --batch-size-file $(realpath benchmarks/dynamo/torchbench_models_list.txt)
```

Explanation of the envvars show up on the command:
```
- TORCHINDUCTOR_COORDINATE_DESCENT_TUNING=1 : enable coordinate descent tuning
- TORCHINDUCTOR_PERSISTENT_REDUCTIONS=0 : disable persistent reduction. Need do this so we can tune RBLOCK for reductions
- TORCHINDUCTOR_MAX_AUTOTUNE=1: enable max autotune
- TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_shunting_coordesc : use a separate cache dir for coordinate descent tuning. Optional.
```

Here are my experiments results for around 40 torchbench models: https://docs.google.com/spreadsheets/d/1G7i2whIf8Yu-HhN_WovNxwcE-iFDSAw4x3NK4uL4XhI/edit#gid=0 

Some highlights
- We improve 2.2% further upon max-autotune on average (geomean)
- timm_resnest benefits most from coordinate descent tuning. There is 1.07x speedup
- We have descent speedup on transformer models
  - BERT_pytorch:  1.056x
  - timm_vision_transformer: 1.04x
  - hf_Bert: 1.030x
- For resnet models, it looks like we have less gain as model get larger. My guess is larger model spend more time on mm/conv, so our tuning for pointwise/reduction helps less
  - resnet18: 1.021x
  - resnet50: 1.014x
  - resnet152: 1.005x

This kind of coordinate descent autotuning can give us 'upper bound' of the gain we can get for tuning configs for pointwise/reduction. On the other hand, by spot checking, we roughly double the compilation time compared to max-autotune. Next steps can be
- we disable persistent reduction in coordinate descent autotune (it's still enabled in baseline) so we can tune RBLOCK for reduction. We can also try to use autotune to pick persistent reduction or not.
- pick good config without benchmarking (e.g. Natalia mentioned checking register spill)
- try the idea on matmul so we know what's the potential there.





cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

shunting314 committed:
[WIP][inductor] coordinate descent tuning upon max-autotune

[ghstack-poisoned]

ngimel commented:
Can you try on current master? Some alignment errors have been fixed in triton, and I've landed new triton pin today

shunting314 committed:
Update on "[WIP][inductor] coordinate descent tuning upon max-autotune"


Some early result for hf_Bert inference:
```
python benchmarks/dynamo/torchbench.py --backend inductor --amp --performance --dashboard --only hf_Bert --disable-cudagraphs

plain autotune: 1.175x

max autotune: 1.205x 

coordinate descent autotune: 1.353x p=0.00
```

Tried it for more cases like hf_Bert training, the autotuning crashes (unaligned access in GPU). I think we should leverage the sub-process tuning.


Still a draft. cc ngimel  Chillee jansel 



[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] coordinate descent tuning upon max-autotune"


Some early result for hf_Bert inference:
```
python benchmarks/dynamo/torchbench.py --backend inductor --amp --performance --dashboard --only hf_Bert --disable-cudagraphs

plain autotune: 1.175x

max autotune: 1.205x 

coordinate descent autotune: 1.353x p=0.00
```

Tried it for more cases like hf_Bert training, the autotuning crashes (unaligned access in GPU). I think we should leverage the sub-process tuning.


Still a draft. cc ngimel  Chillee jansel 



[ghstack-poisoned]

shunting314 commented:
> Can you try on current master? Some alignment errors have been fixed in triton, and I've landed new triton pin today

After rebasing, I seem to pass the place where it fails previously. But now I get 
```
RuntimeError: Triton Error [CUDA]: invalid argument
```
Maybe some config we tried is too crazy. I'll try to print what config it is.

But before that, I think I need add the best choice found by coordinate descent into cache so we don't need start over every time. That can make us iterate much faster.

Chillee commented:
I think it would be ideal if we could just capture the kernels out separately so we don't need to run things model by model. But up to you.

shunting314 commented:
> I think it would be ideal if we could just capture the kernels out separately so we don't need to run things model by model. But up to you.

yea, before trying this on a real model, I actually tried it on the hf_Bert kernel you showed me. The algorithm actually found the same best config we found manually.

shunting314 commented:
> I think it would be ideal if we could just capture the kernels out separately so we don't need to run things model by model. But up to you.

On the other hand, this algorithm assumes the system as a block box. But it would be nice that we capture the kernels that is speed up most and verify how the config change affect the perf. I've see quite a few kernels being optimized more than 5x in hf_Bert training before the crash happens.  

```
= Do coordinate descent tuning =
Tune from XBLOCK: 1, RBLOCK: 512, num_warps: 4, num_stages: 2 0.12492799758911133 -> XBLOCK: 1, RBLOCK: 256, num_warps: 4, num_stages: 2 0.07884799689054489
Tune from XBLOCK: 1, RBLOCK: 256, num_warps: 4, num_stages: 2 0.07884799689054489 -> XBLOCK: 1, RBLOCK: 256, num_warps: 2, num_stages: 2 0.06860800087451935
Tune from XBLOCK: 1, RBLOCK: 256, num_warps: 2, num_stages: 2 0.06860800087451935 -> XBLOCK: 1, RBLOCK: 128, num_warps: 2, num_stages: 2 0.04403200000524521
Tune from XBLOCK: 1, RBLOCK: 128, num_warps: 2, num_stages: 2 0.04403200000524521 -> XBLOCK: 1, RBLOCK: 128, num_warps: 1, num_stages: 2 0.03891199827194214
Tune from XBLOCK: 1, RBLOCK: 128, num_warps: 1, num_stages: 2 0.03891199827194214 -> XBLOCK: 1, RBLOCK: 64, num_warps: 1, num_stages: 2 0.03379200026392937
Tune from XBLOCK: 1, RBLOCK: 64, num_warps: 1, num_stages: 2 0.03379200026392937 -> XBLOCK: 2, RBLOCK: 64, num_warps: 1, num_stages: 2 0.025599999353289604
Tune from XBLOCK: 2, RBLOCK: 64, num_warps: 1, num_stages: 2 0.025599999353289604 -> XBLOCK: 2, RBLOCK: 32, num_warps: 1, num_stages: 2 0.02252800017595291
Tune from XBLOCK: 2, RBLOCK: 32, num_warps: 1, num_stages: 2 0.02252800017595291 -> XBLOCK: 4, RBLOCK: 32, num_warps: 1, num_stages: 2 0.01945599913597107
Tune from XBLOCK: 4, RBLOCK: 32, num_warps: 1, num_stages: 2 0.01945599913597107 -> XBLOCK: 4, RBLOCK: 16, num_warps: 1, num_stages: 2 0.018432000651955605
Tune from XBLOCK: 4, RBLOCK: 16, num_warps: 1, num_stages: 2 0.018432000651955605 -> XBLOCK: 8, RBLOCK: 16, num_warps: 1, num_stages: 2 0.016383999958634377
```

shunting314 committed:
Update on "[WIP][inductor] coordinate descent tuning upon max-autotune"


Some early result for hf_Bert inference:
```
python benchmarks/dynamo/torchbench.py --backend inductor --amp --performance --dashboard --only hf_Bert --disable-cudagraphs

plain autotune: 1.175x

max autotune: 1.205x 

coordinate descent autotune: 1.353x p=0.00
```

Tried it for more cases like hf_Bert training, the autotuning crashes (unaligned access in GPU). I think we should leverage the sub-process tuning.


Still a draft. cc ngimel  Chillee jansel 



[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] coordinate descent tuning upon max-autotune"


Some early result for hf_Bert inference:
```
python benchmarks/dynamo/torchbench.py --backend inductor --amp --performance --dashboard --only hf_Bert --disable-cudagraphs

plain autotune: 1.175x

max autotune: 1.205x 

coordinate descent autotune: 1.353x p=0.00
```

Tried it for more cases like hf_Bert training, the autotuning crashes (unaligned access in GPU). I think we should leverage the sub-process tuning.


Still a draft. cc ngimel  Chillee jansel 



[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] coordinate descent tuning upon max-autotune"


Previous results tunes RBLOCK for persistent reduction, which should be invalid. Will rerun experiments and report new numbers.


Still a draft. cc ngimel  Chillee jansel 



[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] coordinate descent tuning upon max-autotune"


Previous results tunes RBLOCK for persistent reduction, which should be invalid. Will rerun experiments and report new numbers.


Still a draft. cc ngimel  Chillee jansel 



[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] coordinate descent tuning upon max-autotune"


Previous results tunes RBLOCK for persistent reduction, which should be invalid. Will rerun experiments and report new numbers.


Still a draft. cc ngimel  Chillee jansel 



[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] coordinate descent tuning upon max-autotune"


Previous results tunes RBLOCK for persistent reduction, which should be invalid. Will rerun experiments and report new numbers.


Still a draft. cc ngimel  Chillee jansel 



[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] coordinate descent tuning upon max-autotune"


Previous results tunes RBLOCK for persistent reduction, which should be invalid. Will rerun experiments and report new numbers.


Still a draft. cc ngimel  Chillee jansel 



[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] coordinate descent tuning upon max-autotune"


Previous results tunes RBLOCK for persistent reduction, which should be invalid. Will rerun experiments and report new numbers.


Still a draft. cc ngimel  Chillee jansel 



[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] coordinate descent tuning upon max-autotune"


Previous results tunes RBLOCK for persistent reduction, which should be invalid. Will rerun experiments and report new numbers.


Still a draft. cc ngimel  Chillee jansel 



[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] coordinate descent tuning upon max-autotune"


Previous results tunes RBLOCK for persistent reduction, which should be invalid. Will rerun experiments and report new numbers.


Still a draft. cc ngimel  Chillee jansel 



[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] coordinate descent tuning upon max-autotune"


Previous results tunes RBLOCK for persistent reduction, which should be invalid. Will rerun experiments and report new numbers.


Still a draft. cc ngimel  Chillee jansel 



[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] coordinate descent tuning upon max-autotune"


Previous results tunes RBLOCK for persistent reduction, which should be invalid. Will rerun experiments and report new numbers.


Still a draft. cc ngimel  Chillee jansel 



[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] coordinate descent tuning upon max-autotune"


Previous results tunes RBLOCK for persistent reduction, which should be invalid. Will rerun experiments and report new numbers.


Still a draft. cc ngimel  Chillee jansel 



[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] coordinate descent tuning upon max-autotune"


Previous results tunes RBLOCK for persistent reduction, which should be invalid. Will rerun experiments and report new numbers.


Still a draft. cc ngimel  Chillee jansel 



[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] coordinate descent tuning upon max-autotune"


Previous results tunes RBLOCK for persistent reduction, which should be invalid. Will rerun experiments and report new numbers.


Still a draft. cc ngimel  Chillee jansel 



[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] coordinate descent tuning upon max-autotune"


Previous results tunes RBLOCK for persistent reduction, which should be invalid. Will rerun experiments and report new numbers.


Still a draft. cc ngimel  Chillee jansel 



[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] coordinate descent tuning upon max-autotune"


Previous results tunes RBLOCK for persistent reduction, which should be invalid. Will rerun experiments and report new numbers.


Still a draft. cc ngimel  Chillee jansel 



[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] coordinate descent tuning upon max-autotune"


Previous results tunes RBLOCK for persistent reduction, which should be invalid. Will rerun experiments and report new numbers.


Still a draft. cc ngimel  Chillee jansel 



[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] coordinate descent tuning upon max-autotune"


Previous results tunes RBLOCK for persistent reduction, which should be invalid. Will rerun experiments and report new numbers.


Still a draft. cc ngimel  Chillee jansel 



[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] coordinate descent tuning upon max-autotune"


Previous results tunes RBLOCK for persistent reduction, which should be invalid. Will rerun experiments and report new numbers.


Still a draft. cc ngimel  Chillee jansel 



[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] coordinate descent tuning upon max-autotune"


Previous results tunes RBLOCK for persistent reduction, which should be invalid. Will rerun experiments and report new numbers.


Still a draft. cc ngimel  Chillee jansel 



[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] coordinate descent tuning upon max-autotune"


Previous results tunes RBLOCK for persistent reduction, which should be invalid. Will rerun experiments and report new numbers.


Still a draft. cc ngimel  Chillee jansel 



[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] coordinate descent tuning upon max-autotune"


Previous results tunes RBLOCK for persistent reduction, which should be invalid. Will rerun experiments and report new numbers.


Still a draft. cc ngimel  Chillee jansel 



[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] coordinate descent tuning upon max-autotune"


Previous results tunes RBLOCK for persistent reduction, which should be invalid. Will rerun experiments and report new numbers.


Still a draft. cc ngimel  Chillee jansel 



[ghstack-poisoned]

shunting314 committed:
Update on "[WIP][inductor] coordinate descent tuning upon max-autotune"


Previous results tunes RBLOCK for persistent reduction, which should be invalid. Will rerun experiments and report new numbers.


Still a draft. cc ngimel  Chillee jansel 



[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] coordinate descent tuning upon max-autotune"


Command to run max autotune baseline:
```
TORCHINDUCTOR_MAX_AUTOTUNE=1 time python benchmarks/dynamo/torchbench.py --backend inductor --amp --performance --only ${MODEL_NAME} --training --batch-size-file $(realpath benchmarks/dynamo/torchbench_models_list.txt)
```

Command to do coordinate descent autotuning: 
```
TORCHINDUCTOR_COORDINATE_DESCENT_TUNING=1 TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_shunting_coordesc TORCHINDUCTOR_PERSISTENT_REDUCTIONS=0 TORCHINDUCTOR_MAX_AUTOTUNE=1 time python benchmarks/dynamo/torchbench.py --backend inductor --amp --performance --only ${MODEL_NAME} --training --batch-size-file $(realpath benchmarks/dynamo/torchbench_models_list.txt)
```

Some results:
timm_vision_transformer
- baseline: 98.8ms
- coordesc tuner: 94.8685ms , 1.041x speedup





[ghstack-poisoned]

Chillee commented on torch/_inductor/triton_heuristics.py:
```diff
@@ -67,6 +84,8 @@ def __init__(self, fn, meta, configs, save_cache_hook, mutated_arg_names):
                 str(self.meta.get("device", 0)),
             )
 
+        self.coordesc_tuner = CoordescTuner(is_mm=False, is_persistent_reduction=self.is_persistent_reduction())
```
btw, I think we should just modify codegen for persistent reductions so that RBLOCK isn't a kernel parameter.

shunting314 replied:
We also need remove RBLOCK from persistent reduction's triton.Config? Since the coordinate descent tuning looks for all the fields in a triton.Config and try to tune them.

Chillee replied:
Yeah we should just make the changes to use rnumel instead of RBLOCK and not make it a tunable parameter in the config.

shunting314 replied:
I just found RBLOCK is a power of 2 while rnumel may be not. E.g.
```
@triton.jit
def triton_(in_out_ptr0, in_ptr0, in_ptr1, in_ptr2, in_ptr3, in_ptr4, in_ptr5, in_ptr6, in_ptr7, out_ptr2, xnumel, rnumel, XBLOCK : tl.constexpr, RBLOCK : tl.constexpr):
    xnumel = 4096
    rnumel = 768
```

Should we be worried about that? Not sure if this impact perf.

shunting314 replied:
To be safe, should we generated code like:
```
@persistent_reduction(
    size_hints=[4096, 1024],
    reduction_hint=ReductionHint.INNER,
    filename=__file__,
    meta={'signature': {0: '*fp32', 1: '*i64', 2: '*fp16', 3: '*i64', 4: '*fp16', 5: '*i64', 6: '*fp16', 7: '*fp16', 8: '*fp16', 9: '*fp16', 10: 'i32', 11: 'i32'}, 'device': 0, 'constants': {}, 'mutated_arg_names': ['in_out_ptr0'], 'configs': [instance_descriptor(divisible_by_16=(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11), equal_to_1=())]}
)
@triton.jit
def triton_(in_out_ptr0, in_ptr0, in_ptr1, in_ptr2, in_ptr3, in_ptr4, in_ptr5, in_ptr6, in_ptr7, out_ptr2, xnumel, rnumel, XBLOCK : tl.constexpr):
    xnumel = 4096
    rnumel = 768
    RBLOCK = 1024 # round rnumel to the next power of 2
```

cc @Chillee

ngimel replied:
Yes for persistent reduction we should codegen RBLOCK that's the next power of 2.

shunting314 replied:
I resolve this comment in a separate PR here https://github.com/pytorch/pytorch/pull/98653 .

shunting314 committed:
Update on "[inductor] coordinate descent tuning upon max-autotune"


Command to run max autotune baseline:
```
TORCHINDUCTOR_MAX_AUTOTUNE=1 time python benchmarks/dynamo/torchbench.py --backend inductor --amp --performance --only ${MODEL_NAME} --training --batch-size-file $(realpath benchmarks/dynamo/torchbench_models_list.txt)
```

Command to do coordinate descent autotuning: 
```
TORCHINDUCTOR_COORDINATE_DESCENT_TUNING=1 TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_shunting_coordesc TORCHINDUCTOR_PERSISTENT_REDUCTIONS=0 TORCHINDUCTOR_MAX_AUTOTUNE=1 time python benchmarks/dynamo/torchbench.py --backend inductor --amp --performance --only ${MODEL_NAME} --training --batch-size-file $(realpath benchmarks/dynamo/torchbench_models_list.txt)
```

Here are my experiments results for around 40 torchbench models: https://docs.google.com/spreadsheets/d/1G7i2whIf8Yu-HhN_WovNxwcE-iFDSAw4x3NK4uL4XhI/edit#gid=0 

Some highlights
- We improve 2.2% further upon max-autotune on average (geomean)
- timm_resnest benefits most from coordinate descent tuning. There is 1.07x speedup
- We have descent speedup on transformer models
  - BERT_pytorch:  1.056x
  - timm_vision_transformer: 1.04x
  - hf_Bert: 1.030x
- For resnet models, it looks like we have less gain as model get larger. My guess is larger model spend more time on mm/conv, so our tuning for pointwise/reduction helps less
  - resnet18: 1.021x
  - resnet50: 1.014x
  - resnet152: 1.005x

This kind of coordinate descent autotuning can give us 'upper bound' of the gain we can get for tuning configs for pointwise/reduction. On the other hand, by spot checking, we roughly double the compilation time compared to max-autotune. Next steps can be
- we disable persistent reduction in coordinate descent autotune (it's still enabled in baseline) so we can tune RBLOCK for reduction. We can also try to use autotune to pick persistent reduction or not.
- pick good config without benchmarking (e.g. Natalia mentioned checking register spill)
- try the idea on matmul so we know what's the potential there.





[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] coordinate descent tuning upon max-autotune"


Command to run max autotune baseline:
```
TORCHINDUCTOR_MAX_AUTOTUNE=1 time python benchmarks/dynamo/torchbench.py --backend inductor --amp --performance --only ${MODEL_NAME} --training --batch-size-file $(realpath benchmarks/dynamo/torchbench_models_list.txt)
```

Command to do coordinate descent autotuning: 
```
TORCHINDUCTOR_COORDINATE_DESCENT_TUNING=1 TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_shunting_coordesc TORCHINDUCTOR_PERSISTENT_REDUCTIONS=0 TORCHINDUCTOR_MAX_AUTOTUNE=1 time python benchmarks/dynamo/torchbench.py --backend inductor --amp --performance --only ${MODEL_NAME} --training --batch-size-file $(realpath benchmarks/dynamo/torchbench_models_list.txt)
```

Here are my experiments results for around 40 torchbench models: https://docs.google.com/spreadsheets/d/1G7i2whIf8Yu-HhN_WovNxwcE-iFDSAw4x3NK4uL4XhI/edit#gid=0 

Some highlights
- We improve 2.2% further upon max-autotune on average (geomean)
- timm_resnest benefits most from coordinate descent tuning. There is 1.07x speedup
- We have descent speedup on transformer models
  - BERT_pytorch:  1.056x
  - timm_vision_transformer: 1.04x
  - hf_Bert: 1.030x
- For resnet models, it looks like we have less gain as model get larger. My guess is larger model spend more time on mm/conv, so our tuning for pointwise/reduction helps less
  - resnet18: 1.021x
  - resnet50: 1.014x
  - resnet152: 1.005x

This kind of coordinate descent autotuning can give us 'upper bound' of the gain we can get for tuning configs for pointwise/reduction. On the other hand, by spot checking, we roughly double the compilation time compared to max-autotune. Next steps can be
- we disable persistent reduction in coordinate descent autotune (it's still enabled in baseline) so we can tune RBLOCK for reduction. We can also try to use autotune to pick persistent reduction or not.
- pick good config without benchmarking (e.g. Natalia mentioned checking register spill)
- try the idea on matmul so we know what's the potential there.





[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] coordinate descent tuning upon max-autotune"


Command to run max autotune baseline:
```
TORCHINDUCTOR_MAX_AUTOTUNE=1 time python benchmarks/dynamo/torchbench.py --backend inductor --amp --performance --only ${MODEL_NAME} --training --batch-size-file $(realpath benchmarks/dynamo/torchbench_models_list.txt)
```

Command to do coordinate descent autotuning: 
```
TORCHINDUCTOR_COORDINATE_DESCENT_TUNING=1 TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_shunting_coordesc TORCHINDUCTOR_PERSISTENT_REDUCTIONS=0 TORCHINDUCTOR_MAX_AUTOTUNE=1 time python benchmarks/dynamo/torchbench.py --backend inductor --amp --performance --only ${MODEL_NAME} --training --batch-size-file $(realpath benchmarks/dynamo/torchbench_models_list.txt)
```

Explanation of the envvars show up on the command:
```
- TORCHINDUCTOR_COORDINATE_DESCENT_TUNING=1 : enable coordinate descent tuning
- TORCHINDUCTOR_PERSISTENT_REDUCTIONS=0 : disable persistent reduction. Need do this so we can tune RBLOCK for reductions
- TORCHINDUCTOR_MAX_AUTOTUNE=1: enable max autotune
- TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_shunting_coordesc : use a separate cache dir for coordinate descent tuning. Optional.
```

Here are my experiments results for around 40 torchbench models: https://docs.google.com/spreadsheets/d/1G7i2whIf8Yu-HhN_WovNxwcE-iFDSAw4x3NK4uL4XhI/edit#gid=0 

Some highlights
- We improve 2.2% further upon max-autotune on average (geomean)
- timm_resnest benefits most from coordinate descent tuning. There is 1.07x speedup
- We have descent speedup on transformer models
  - BERT_pytorch:  1.056x
  - timm_vision_transformer: 1.04x
  - hf_Bert: 1.030x
- For resnet models, it looks like we have less gain as model get larger. My guess is larger model spend more time on mm/conv, so our tuning for pointwise/reduction helps less
  - resnet18: 1.021x
  - resnet50: 1.014x
  - resnet152: 1.005x

This kind of coordinate descent autotuning can give us 'upper bound' of the gain we can get for tuning configs for pointwise/reduction. On the other hand, by spot checking, we roughly double the compilation time compared to max-autotune. Next steps can be
- we disable persistent reduction in coordinate descent autotune (it's still enabled in baseline) so we can tune RBLOCK for reduction. We can also try to use autotune to pick persistent reduction or not.
- pick good config without benchmarking (e.g. Natalia mentioned checking register spill)
- try the idea on matmul so we know what's the potential there.





[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] coordinate descent tuning upon max-autotune"


Command to run max autotune baseline:
```
TORCHINDUCTOR_MAX_AUTOTUNE=1 time python benchmarks/dynamo/torchbench.py --backend inductor --amp --performance --only ${MODEL_NAME} --training --batch-size-file $(realpath benchmarks/dynamo/torchbench_models_list.txt)
```

Command to do coordinate descent autotuning: 
```
TORCHINDUCTOR_COORDINATE_DESCENT_TUNING=1 TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_shunting_coordesc TORCHINDUCTOR_PERSISTENT_REDUCTIONS=0 TORCHINDUCTOR_MAX_AUTOTUNE=1 time python benchmarks/dynamo/torchbench.py --backend inductor --amp --performance --only ${MODEL_NAME} --training --batch-size-file $(realpath benchmarks/dynamo/torchbench_models_list.txt)
```

Explanation of the envvars show up on the command:
```
- TORCHINDUCTOR_COORDINATE_DESCENT_TUNING=1 : enable coordinate descent tuning
- TORCHINDUCTOR_PERSISTENT_REDUCTIONS=0 : disable persistent reduction. Need do this so we can tune RBLOCK for reductions
- TORCHINDUCTOR_MAX_AUTOTUNE=1: enable max autotune
- TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_shunting_coordesc : use a separate cache dir for coordinate descent tuning. Optional.
```

Here are my experiments results for around 40 torchbench models: https://docs.google.com/spreadsheets/d/1G7i2whIf8Yu-HhN_WovNxwcE-iFDSAw4x3NK4uL4XhI/edit#gid=0 

Some highlights
- We improve 2.2% further upon max-autotune on average (geomean)
- timm_resnest benefits most from coordinate descent tuning. There is 1.07x speedup
- We have descent speedup on transformer models
  - BERT_pytorch:  1.056x
  - timm_vision_transformer: 1.04x
  - hf_Bert: 1.030x
- For resnet models, it looks like we have less gain as model get larger. My guess is larger model spend more time on mm/conv, so our tuning for pointwise/reduction helps less
  - resnet18: 1.021x
  - resnet50: 1.014x
  - resnet152: 1.005x

This kind of coordinate descent autotuning can give us 'upper bound' of the gain we can get for tuning configs for pointwise/reduction. On the other hand, by spot checking, we roughly double the compilation time compared to max-autotune. Next steps can be
- we disable persistent reduction in coordinate descent autotune (it's still enabled in baseline) so we can tune RBLOCK for reduction. We can also try to use autotune to pick persistent reduction or not.
- pick good config without benchmarking (e.g. Natalia mentioned checking register spill)
- try the idea on matmul so we know what's the potential there.





[ghstack-poisoned]

ngimel commented on torch/_inductor/triton_heuristics.py:
```diff
@@ -51,13 +63,22 @@ class CachingAutotuner(KernelInterface):
         self.configs = configs
+        self.heuristic_type = heuristic_type
+
+        if DEBUG:
```
can you make it log.debug?

ngimel commented on torch/_inductor/triton_heuristics.py:
```diff
@@ -352,22 +433,30 @@ def cached_autotune(
+                        }
+                    )
+                )
+            if DEBUG:
```
same, log.debug

ngimel commented on torch/_inductor/triton_heuristics.py:
```diff
@@ -605,14 +714,32 @@ def reduction(size_hints, reduction_hint=False, meta=None, filename=None):
+                [contiguous_config],
+                meta=meta,
+                heuristic_type=HeuristicType.REDUCTION,
+                filename=filename,
```
why did we not need to pass filename previously?

shunting314 replied:
So previously if there is a single config passed in, we don't need do any tuning. But now even if there is a single config, we may still want to do the coordinate descent tuning. We want to write the tuning results to the cache.

The filename is passed in since the cache path is computed based on filename.

ngimel reviewed (commented):
The way PR is structured it's a bit hard to make sure that this doesn't change behavior and caching when coordinate descent is not enabled. Can you double check that caching for no-autotuning/no coordinate descent still works as before?

ngimel commented:
Can you describe what happens when 
1) both max_autotune and coordinate_descent are on
2) max_autotune is off, but coordinate_descent is on

shunting314 commented:
yep.

> 1. both max_autotune and coordinate_descent are on
> 2. max_autotune is off, but coordinate_descent is on

for the above 2 scenarios, the only difference is the starting config for coordinate_descent tuning.

E.g., assuming regular autotune only get one config C1; while max-autotune get 4 configs C1, C2, C3, C4 and max-autotune figure out C3 is the best.

Then in scenario 1 you mentioned. coordinate descent tuning will start from C1;
while in scenario 2, coordinate descent tuning will start from C3

ngimel commented:
Great, can you leave this comment in `cached_autotune`?

shunting314 commented:
> The way PR is structured it's a bit hard to make sure that this doesn't change behavior and caching when coordinate descent is not enabled. Can you double check that caching for no-autotuning/no coordinate descent still works as before?

Testing on a specific kernel is much easier compared to testing on a whole model. So I tested on this kernel: https://gist.github.com/shunting314/80432018df55aab48444429ad971e176  (named k.py)
- out of box behavior

run `python k.py`. Cache is skipped since this kernel only has 1 config by default.

- max autotune

run `TORCHINDUCTOR_MAX_AUTOTUNE=1 python /tmp/k.py`, the cache metadata file contains:
```
{"XBLOCK": 16, "YBLOCK": 256, "num_warps": 8, "num_stages": 1, "configs_hash": "9c03eb9a92d42ada89eb81e670a49ca46062f487a8f33409913c8e62525a06ca", "found_by_coordesc": false}
```
It contains a config_hash for all the configs considered and the metadata for the picked config. It also tells this best config is not found by coordinate descent search.

- run coordinate descent tuning upon maxautotune

run `TORCHINDUCTOR_COORDINATE_DESCENT_TUNING=1 TORCHINDUCTOR_MAX_AUTOTUNE=1 python /tmp/k.py`, the cache metadata file contains:

```
{"XBLOCK": 8, "YBLOCK": 512, "num_warps": 16, "num_stages": 1, "configs_hash": "9c03eb9a92d42ada89eb81e670a49ca46062f487a8f33409913c8e62525a06ca", "found_by_coordesc": true}
```

We can see that `configs_hash` is the same as before since we are based on the same max-autotune configs. We see found_by_coordesc is true since we have run coordinate descent tuning. We see coordinate descent tuning find a better config (1.127x faster)

- run maxautotune without coordinate descent tuning again

The cache metadata recovers to 
```
{"XBLOCK": 16, "YBLOCK": 256, "num_warps": 8, "num_stages": 1, "configs_hash": "9c03eb9a92d42ada89eb81e670a49ca46062f487a8f33409913c8e62525a06ca", "found_by_coordesc": false}
```
Note that the config found by coordinate descent tuning previously is dropped since it does not match any of the candidate config that max-autotune check.


In my tests, I usually use a different inductor cache directory when testing coordinate descent tuning to avoid the ping-pong effect:
- running coordinate descent tuning overwrite the warmed up max-autotune cache
- running max-autotune overwrite the warmed up coordinate descent tuning cache

shunting314 committed:
Update on "[inductor] coordinate descent tuning upon max-autotune"


Command to run max autotune baseline:
```
TORCHINDUCTOR_MAX_AUTOTUNE=1 time python benchmarks/dynamo/torchbench.py --backend inductor --amp --performance --only ${MODEL_NAME} --training --batch-size-file $(realpath benchmarks/dynamo/torchbench_models_list.txt)
```

Command to do coordinate descent autotuning: 
```
TORCHINDUCTOR_COORDINATE_DESCENT_TUNING=1 TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_shunting_coordesc TORCHINDUCTOR_PERSISTENT_REDUCTIONS=0 TORCHINDUCTOR_MAX_AUTOTUNE=1 time python benchmarks/dynamo/torchbench.py --backend inductor --amp --performance --only ${MODEL_NAME} --training --batch-size-file $(realpath benchmarks/dynamo/torchbench_models_list.txt)
```

Explanation of the envvars show up on the command:
```
- TORCHINDUCTOR_COORDINATE_DESCENT_TUNING=1 : enable coordinate descent tuning
- TORCHINDUCTOR_PERSISTENT_REDUCTIONS=0 : disable persistent reduction. Need do this so we can tune RBLOCK for reductions
- TORCHINDUCTOR_MAX_AUTOTUNE=1: enable max autotune
- TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_shunting_coordesc : use a separate cache dir for coordinate descent tuning. Optional.
```

Here are my experiments results for around 40 torchbench models: https://docs.google.com/spreadsheets/d/1G7i2whIf8Yu-HhN_WovNxwcE-iFDSAw4x3NK4uL4XhI/edit#gid=0 

Some highlights
- We improve 2.2% further upon max-autotune on average (geomean)
- timm_resnest benefits most from coordinate descent tuning. There is 1.07x speedup
- We have descent speedup on transformer models
  - BERT_pytorch:  1.056x
  - timm_vision_transformer: 1.04x
  - hf_Bert: 1.030x
- For resnet models, it looks like we have less gain as model get larger. My guess is larger model spend more time on mm/conv, so our tuning for pointwise/reduction helps less
  - resnet18: 1.021x
  - resnet50: 1.014x
  - resnet152: 1.005x

This kind of coordinate descent autotuning can give us 'upper bound' of the gain we can get for tuning configs for pointwise/reduction. On the other hand, by spot checking, we roughly double the compilation time compared to max-autotune. Next steps can be
- we disable persistent reduction in coordinate descent autotune (it's still enabled in baseline) so we can tune RBLOCK for reduction. We can also try to use autotune to pick persistent reduction or not.
- pick good config without benchmarking (e.g. Natalia mentioned checking register spill)
- try the idea on matmul so we know what's the potential there.





cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] coordinate descent tuning upon max-autotune"


Command to run max autotune baseline:
```
TORCHINDUCTOR_MAX_AUTOTUNE=1 time python benchmarks/dynamo/torchbench.py --backend inductor --amp --performance --only ${MODEL_NAME} --training --batch-size-file $(realpath benchmarks/dynamo/torchbench_models_list.txt)
```

Command to do coordinate descent autotuning: 
```
TORCHINDUCTOR_COORDINATE_DESCENT_TUNING=1 TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_shunting_coordesc TORCHINDUCTOR_PERSISTENT_REDUCTIONS=0 TORCHINDUCTOR_MAX_AUTOTUNE=1 time python benchmarks/dynamo/torchbench.py --backend inductor --amp --performance --only ${MODEL_NAME} --training --batch-size-file $(realpath benchmarks/dynamo/torchbench_models_list.txt)
```

Explanation of the envvars show up on the command:
```
- TORCHINDUCTOR_COORDINATE_DESCENT_TUNING=1 : enable coordinate descent tuning
- TORCHINDUCTOR_PERSISTENT_REDUCTIONS=0 : disable persistent reduction. Need do this so we can tune RBLOCK for reductions
- TORCHINDUCTOR_MAX_AUTOTUNE=1: enable max autotune
- TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_shunting_coordesc : use a separate cache dir for coordinate descent tuning. Optional.
```

Here are my experiments results for around 40 torchbench models: https://docs.google.com/spreadsheets/d/1G7i2whIf8Yu-HhN_WovNxwcE-iFDSAw4x3NK4uL4XhI/edit#gid=0 

Some highlights
- We improve 2.2% further upon max-autotune on average (geomean)
- timm_resnest benefits most from coordinate descent tuning. There is 1.07x speedup
- We have descent speedup on transformer models
  - BERT_pytorch:  1.056x
  - timm_vision_transformer: 1.04x
  - hf_Bert: 1.030x
- For resnet models, it looks like we have less gain as model get larger. My guess is larger model spend more time on mm/conv, so our tuning for pointwise/reduction helps less
  - resnet18: 1.021x
  - resnet50: 1.014x
  - resnet152: 1.005x

This kind of coordinate descent autotuning can give us 'upper bound' of the gain we can get for tuning configs for pointwise/reduction. On the other hand, by spot checking, we roughly double the compilation time compared to max-autotune. Next steps can be
- we disable persistent reduction in coordinate descent autotune (it's still enabled in baseline) so we can tune RBLOCK for reduction. We can also try to use autotune to pick persistent reduction or not.
- pick good config without benchmarking (e.g. Natalia mentioned checking register spill)
- try the idea on matmul so we know what's the potential there.





cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] coordinate descent tuning upon max-autotune"


Command to run max autotune baseline:
```
TORCHINDUCTOR_MAX_AUTOTUNE=1 time python benchmarks/dynamo/torchbench.py --backend inductor --amp --performance --only ${MODEL_NAME} --training --batch-size-file $(realpath benchmarks/dynamo/torchbench_models_list.txt)
```

Command to do coordinate descent autotuning: 
```
TORCHINDUCTOR_COORDINATE_DESCENT_TUNING=1 TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_shunting_coordesc TORCHINDUCTOR_PERSISTENT_REDUCTIONS=0 TORCHINDUCTOR_MAX_AUTOTUNE=1 time python benchmarks/dynamo/torchbench.py --backend inductor --amp --performance --only ${MODEL_NAME} --training --batch-size-file $(realpath benchmarks/dynamo/torchbench_models_list.txt)
```

Explanation of the envvars show up on the command:
```
- TORCHINDUCTOR_COORDINATE_DESCENT_TUNING=1 : enable coordinate descent tuning
- TORCHINDUCTOR_PERSISTENT_REDUCTIONS=0 : disable persistent reduction. Need do this so we can tune RBLOCK for reductions
- TORCHINDUCTOR_MAX_AUTOTUNE=1: enable max autotune
- TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_shunting_coordesc : use a separate cache dir for coordinate descent tuning. Optional.
```

Here are my experiments results for around 40 torchbench models: https://docs.google.com/spreadsheets/d/1G7i2whIf8Yu-HhN_WovNxwcE-iFDSAw4x3NK4uL4XhI/edit#gid=0 

Some highlights
- We improve 2.2% further upon max-autotune on average (geomean)
- timm_resnest benefits most from coordinate descent tuning. There is 1.07x speedup
- We have descent speedup on transformer models
  - BERT_pytorch:  1.056x
  - timm_vision_transformer: 1.04x
  - hf_Bert: 1.030x
- For resnet models, it looks like we have less gain as model get larger. My guess is larger model spend more time on mm/conv, so our tuning for pointwise/reduction helps less
  - resnet18: 1.021x
  - resnet50: 1.014x
  - resnet152: 1.005x

This kind of coordinate descent autotuning can give us 'upper bound' of the gain we can get for tuning configs for pointwise/reduction. On the other hand, by spot checking, we roughly double the compilation time compared to max-autotune. Next steps can be
- we disable persistent reduction in coordinate descent autotune (it's still enabled in baseline) so we can tune RBLOCK for reduction. We can also try to use autotune to pick persistent reduction or not.
- pick good config without benchmarking (e.g. Natalia mentioned checking register spill)
- try the idea on matmul so we know what's the potential there.





cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] coordinate descent tuning upon max-autotune"


Command to run max autotune baseline:
```
TORCHINDUCTOR_MAX_AUTOTUNE=1 time python benchmarks/dynamo/torchbench.py --backend inductor --amp --performance --only ${MODEL_NAME} --training --batch-size-file $(realpath benchmarks/dynamo/torchbench_models_list.txt)
```

Command to do coordinate descent autotuning: 
```
TORCHINDUCTOR_COORDINATE_DESCENT_TUNING=1 TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_shunting_coordesc TORCHINDUCTOR_PERSISTENT_REDUCTIONS=0 TORCHINDUCTOR_MAX_AUTOTUNE=1 time python benchmarks/dynamo/torchbench.py --backend inductor --amp --performance --only ${MODEL_NAME} --training --batch-size-file $(realpath benchmarks/dynamo/torchbench_models_list.txt)
```

Explanation of the envvars show up on the command:
```
- TORCHINDUCTOR_COORDINATE_DESCENT_TUNING=1 : enable coordinate descent tuning
- TORCHINDUCTOR_PERSISTENT_REDUCTIONS=0 : disable persistent reduction. Need do this so we can tune RBLOCK for reductions
- TORCHINDUCTOR_MAX_AUTOTUNE=1: enable max autotune
- TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_shunting_coordesc : use a separate cache dir for coordinate descent tuning. Optional.
```

Here are my experiments results for around 40 torchbench models: https://docs.google.com/spreadsheets/d/1G7i2whIf8Yu-HhN_WovNxwcE-iFDSAw4x3NK4uL4XhI/edit#gid=0 

Some highlights
- We improve 2.2% further upon max-autotune on average (geomean)
- timm_resnest benefits most from coordinate descent tuning. There is 1.07x speedup
- We have descent speedup on transformer models
  - BERT_pytorch:  1.056x
  - timm_vision_transformer: 1.04x
  - hf_Bert: 1.030x
- For resnet models, it looks like we have less gain as model get larger. My guess is larger model spend more time on mm/conv, so our tuning for pointwise/reduction helps less
  - resnet18: 1.021x
  - resnet50: 1.014x
  - resnet152: 1.005x

This kind of coordinate descent autotuning can give us 'upper bound' of the gain we can get for tuning configs for pointwise/reduction. On the other hand, by spot checking, we roughly double the compilation time compared to max-autotune. Next steps can be
- we disable persistent reduction in coordinate descent autotune (it's still enabled in baseline) so we can tune RBLOCK for reduction. We can also try to use autotune to pick persistent reduction or not.
- pick good config without benchmarking (e.g. Natalia mentioned checking register spill)
- try the idea on matmul so we know what's the potential there.





cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] coordinate descent tuning upon max-autotune"


Command to run max autotune baseline:
```
TORCHINDUCTOR_MAX_AUTOTUNE=1 time python benchmarks/dynamo/torchbench.py --backend inductor --amp --performance --only ${MODEL_NAME} --training --batch-size-file $(realpath benchmarks/dynamo/torchbench_models_list.txt)
```

Command to do coordinate descent autotuning: 
```
TORCHINDUCTOR_COORDINATE_DESCENT_TUNING=1 TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_shunting_coordesc TORCHINDUCTOR_PERSISTENT_REDUCTIONS=0 TORCHINDUCTOR_MAX_AUTOTUNE=1 time python benchmarks/dynamo/torchbench.py --backend inductor --amp --performance --only ${MODEL_NAME} --training --batch-size-file $(realpath benchmarks/dynamo/torchbench_models_list.txt)
```

Explanation of the envvars show up on the command:
```
- TORCHINDUCTOR_COORDINATE_DESCENT_TUNING=1 : enable coordinate descent tuning
- TORCHINDUCTOR_PERSISTENT_REDUCTIONS=0 : disable persistent reduction. Need do this so we can tune RBLOCK for reductions
- TORCHINDUCTOR_MAX_AUTOTUNE=1: enable max autotune
- TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_shunting_coordesc : use a separate cache dir for coordinate descent tuning. Optional.
```

Here are my experiments results for around 40 torchbench models: https://docs.google.com/spreadsheets/d/1G7i2whIf8Yu-HhN_WovNxwcE-iFDSAw4x3NK4uL4XhI/edit#gid=0 

Some highlights
- We improve 2.2% further upon max-autotune on average (geomean)
- timm_resnest benefits most from coordinate descent tuning. There is 1.07x speedup
- We have descent speedup on transformer models
  - BERT_pytorch:  1.056x
  - timm_vision_transformer: 1.04x
  - hf_Bert: 1.030x
- For resnet models, it looks like we have less gain as model get larger. My guess is larger model spend more time on mm/conv, so our tuning for pointwise/reduction helps less
  - resnet18: 1.021x
  - resnet50: 1.014x
  - resnet152: 1.005x

This kind of coordinate descent autotuning can give us 'upper bound' of the gain we can get for tuning configs for pointwise/reduction. On the other hand, by spot checking, we roughly double the compilation time compared to max-autotune. Next steps can be
- we disable persistent reduction in coordinate descent autotune (it's still enabled in baseline) so we can tune RBLOCK for reduction. We can also try to use autotune to pick persistent reduction or not.
- pick good config without benchmarking (e.g. Natalia mentioned checking register spill)
- try the idea on matmul so we know what's the potential there.





cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

shunting314 commented:
@pytorchbot merge -f "accuracy check failures are unrelated"

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your PR, but it breaks MacOS test in trunk' -c nosignal

shunting314 commented:
@huydhn  any idea why I don't see the breakage of MacOS test on github CI? I only see 2 test failures due to accuracy check which are unrelated.

shunting314 commented:
Is the MacOS tests triggered after I submit the merge request? I can see the failure now but I didn't see it before I submit merge request.

kit1980 commented on torch/_inductor/coordinate_descent_tuner.py:
```diff
@@ -0,0 +1,165 @@
+import logging
+from typing import Callable, Optional
+
+import triton
```
Guard against triton not present, like in other file?

kit1980 commented:
This PR also broke Windows tests, see https://github.com/pytorch/pytorch/actions/runs/4738152877/jobs/8414630039

```
  File "C:\actions-runner\_work\pytorch\pytorch\build\win_tmp\build\torch\_inductor\coordinate_descent_tuner.py", line 5, in <module>
    import triton
ModuleNotFoundError: No module named 'triton'
```
